src/java.base/share/classes/jdk/internal/crac/JDKSocketResourceBase.java

```diff
@@ -0,0 +1,160 @@
+        return params -> {
+            String family = params.get("family");
+            if (family != null && addr != null) {
+                switch (family.toLowerCase()) {
+                    case "ipv6", "inet6" -> {
+                        if (!(addr instanceof InetSocketAddress inetAddr)) {
+                            return false;
+                        } else if (!(inetAddr.getAddress() instanceof Inet6Address)) {
+                            return false;
+                        }
+                    }
+                    case "ipv4", "inet4" -> {
+                        if (!(addr instanceof InetSocketAddress inetAddr)) {
+                            return false;
+                        } else if (!(inetAddr.getAddress() instanceof Inet4Address)) {
+                            return false;
+                        }
+                    }
+                    case "ip", "inet" -> {
+                        if (!(addr instanceof InetSocketAddress inetAddr)) {
+                            return false;
+                        }
+                    }
+                    case "unix" -> {
+                        if (!(addr instanceof UnixDomainSocketAddress)) {
+                            return false;
+                        }
+                    }
+                    default -> throw new IllegalArgumentException("Unknown family: " + family);
+                }
+            }
+            String cfgAddress = params.get(addressKey);
+            String cfgPort = params.get(portKey);
+            String cfgPath = params.get(pathKey);
+            if (cfgAddress != null || cfgPort != null) {
+                if (!(addr instanceof InetSocketAddress inetAddr)) {
+                    return false;
+                }
+                if (cfgAddress != null && !"*".equals(cfgAddress)) {
+                    try {
+                        if (!InetAddress.getByName(cfgAddress).equals(inetAddr.getAddress())) {
+                            return false;
+                        }
+                    } catch (UnknownHostException e) {
+                        return false;
+                    }
+                }
+                if (cfgPort != null && !"*".equals(cfgPort)) {
+                    return Integer.parseInt(cfgPort) == inetAddr.getPort();
+                }
+                return true;
+            } else if (cfgPath != null) {
+                if (!(addr instanceof UnixDomainSocketAddress unixAddr)) {
+                    return false;
+                }
+                return FileSystems.getDefault().getPathMatcher("glob:" + cfgPath)
+                        .matches(unixAddr.getPath());
+            } else {
+                return true;
+            }
+        };
+    }
+
+    @SuppressWarnings("fallthrough")
+    @Override
+    public void afterRestore(Context<? extends Resource> context) throws Exception {
+        // Don't do anything when we've already failed
+        if (originalFd < 0 || error) {
```

Comment:
The orignalFd seems to be a leftover, it's not clear why is it stored as a part of the state. Should not we just consider `error`?

Reply:
Functionally it's used to mark the socket as invalid, and I've kept the FD number for debugging, but I can remove that...